File: middleware/express/examples/simple-example.js
/**
 * Simple Agent Passport Express Middleware Example
 *
 * This example demonstrates the three main usage patterns:
 * 1. Global policy enforcement
 * 2. Route-specific with explicit agent ID
 * 3. Route-specific with header fallback
 */

const express = require("express");
const {
  agentPassportMiddleware,
  requirePolicy,
  requirePolicyWithContext,
} = require("../dist/middleware/express/src/index");

const app = express();
app.use(express.json());

// ============================================================================
// CONFIGURATION
// ============================================================================

const AGENT_ID = "ap_a2d10232c6534523812423eec8a1425c45678"; // Your agent ID

// ============================================================================
// PATTERN 1: GLOBAL POLICY ENFORCEMENT
// ============================================================================

/**
 * Global middleware enforces a specific policy on all routes.
 * Agent ID is extracted from X-Agent-Passport-Id header.
 */
app.use(
  agentPassportMiddleware({
    policyId: "finance.payment.refund.v1", // Enforces refunds policy globally
    failClosed: true,
  })
);

// All routes below now require finance.payment.refund.v1 policy compliance
app.post("/api/refunds", (req, res) => {
  // Policy already verified - safe to process
  const { amount, currency, order_id } = req.body;

  res.json({
    success: true,
    refund_id: `ref_${Date.now()}`,
    amount,
    currency,
    order_id,
    agent_id: req.agent.agent_id,
  });
});

// ============================================================================
// PATTERN 2: ROUTE-SPECIFIC WITH EXPLICIT AGENT ID (PREFERRED)
// ============================================================================

/**
 * Explicit agent ID is most secure and clear.
 * No header extraction needed.
 */
app.post(
  "/api/data/export",
  requirePolicy("data.export.create.v1", AGENT_ID),
  (req, res) => {
    // Policy verified with explicit agent ID
    const { rows, format, contains_pii } = req.body;

    res.json({
      success: true,
      export_id: `exp_${Date.now()}`,
      rows,
      format,
      contains_pii,
      agent_id: req.agent.agent_id,
    });
  }
);

// ============================================================================
// PATTERN 3: ROUTE-SPECIFIC WITH HEADER FALLBACK
// ============================================================================

/**
 * Header fallback for backward compatibility.
 * Uses X-Agent-Passport-Id header.
 */
app.post(
  "/api/messages/send",
  requirePolicy("messaging.message.send.v1"), // No agent ID - uses header
  (req, res) => {
    // Policy verified via header
    const { channel, message_count, mentions } = req.body;

    res.json({
      success: true,
      message_id: `msg_${Date.now()}`,
      channel,
      message_count,
      mentions,
      agent_id: req.agent.agent_id,
    });
  }
);

// ============================================================================
// PATTERN 4: CUSTOM CONTEXT
// ============================================================================

/**
 * Custom context for complex scenarios.
 */
app.post(
  "/api/repo/pr",
  requirePolicyWithContext(
    "code.repository.merge.v1",
    {
      repository: "myorg/myrepo",
      base_branch: "main",
    },
    AGENT_ID
  ),
  (req, res) => {
    // Policy verified with custom context
    const { pr_size_kb, file_path } = req.body;

    res.json({
      success: true,
      pr_id: `pr_${Date.now()}`,
      repository: "myorg/myrepo",
      base_branch: "main",
      pr_size_kb,
      file_path,
      agent_id: req.agent.agent_id,
    });
  }
);

// ============================================================================
// ERROR HANDLING
// ============================================================================

app.use((err, req, res, next) => {
  console.error("Error:", err);

  if (err.code === "policy_violation") {
    return res.status(403).json({
      error: "policy_violation",
      message: err.message,
      agent_id: err.agentId,
      policy_id: req.policyId,
    });
  }

  res.status(500).json({
    error: "internal_error",
    message: "Internal server error",
  });
});

// ============================================================================
// HEALTH CHECK
// ============================================================================

app.get("/health", (req, res) => {
  res.json({
    status: "ok",
    timestamp: new Date().toISOString(),
    agent_id: AGENT_ID,
  });
});

// ============================================================================
// STARTUP
// ============================================================================

const PORT = process.env.PORT || 3000;

app.listen(PORT, () => {
  console.log(`🚀 Agent Passport Example Server running on port ${PORT}`);
  console.log(`📋 Agent ID: ${AGENT_ID}`);
  console.log("\n📋 Test your endpoints:");
  console.log("\n1. Refunds (Global Policy):");
  console.log(`curl -X POST 'http://localhost:${PORT}/api/refunds' \\`);
  console.log(`  -H 'Content-Type: application/json' \\`);
  console.log(`  -H 'X-Agent-Passport-Id: ${AGENT_ID}' \\`);
  console.log(
    `  -d '{"amount": 25.00, "currency": "USD", "order_id": "order_123", "customer_id": "cust_456", "reason_code": "defective", "idempotency_key": "idem_789"}'`
  );

  console.log("\n2. Data Export (Explicit Agent ID):");
  console.log(`curl -X POST 'http://localhost:${PORT}/api/data/export' \\`);
  console.log(`  -H 'Content-Type: application/json' \\`);
  console.log(`  -d '{"rows": 1000, "format": "json", "contains_pii": false}'`);

  console.log("\n3. Messaging (Header Fallback):");
  console.log(`curl -X POST 'http://localhost:${PORT}/api/messages/send' \\`);
  console.log(`  -H 'Content-Type: application/json' \\`);
  console.log(`  -H 'X-Agent-Passport-Id: ${AGENT_ID}' \\`);
  console.log(
    `  -d '{"channel": "general", "message_count": 5, "mentions": ["@user1"]}'`
  );

  console.log("\n4. Repository PR (Custom Context):");
  console.log(`curl -X POST 'http://localhost:${PORT}/api/repo/pr' \\`);
  console.log(`  -H 'Content-Type: application/json' \\`);
  console.log(`  -d '{"pr_size_kb": 50, "file_path": "src/main.js"}'`);
});

// ============================================================================
// SUMMARY: SIMPLE PATTERNS
// ============================================================================

/*
KEY PATTERNS:

✅ GLOBAL: Enforce one policy on all routes
   app.use(agentPassportMiddleware({ policyId: "finance.payment.refund.v1" }));

✅ EXPLICIT: Most secure, explicit agent ID
   app.post("/api/refunds", requirePolicy("finance.payment.refund.v1", AGENT_ID), handler);

✅ HEADER: Backward compatible, uses header
   app.post("/api/export", requirePolicy("data.export.create.v1"), handler);

✅ CONTEXT: Custom context for complex scenarios
   app.post("/api/pr", requirePolicyWithContext("code.repository.merge.v1", context, AGENT_ID), handler);

THAT'S IT! Simple, clear, and powerful policy enforcement.
*/;

File: middleware/fastapi/examples/simple_example.py
"""
Simple Agent Passport FastAPI Middleware Example

This example demonstrates the three main usage patterns:
1. Global policy enforcement
2. Route-specific with explicit agent ID
3. Route-specific with header fallback
"""

from fastapi import Depends, FastAPI, Request, HTTPException
from fastapi.responses import JSONResponse
import time
import uvicorn

from aporthq_middleware_fastapi import (
    agent_passport_middleware,
    require_policy,
    require_policy_dependency_with_context,
    AgentPassportMiddlewareOptions,
)

app = FastAPI(title="Agent Passport Example API")

# ============================================================================
# CONFIGURATION
# ============================================================================

AGENT_ID = "ap_a2d10232c6534523812423eec8a1425c45678"  # Your agent ID

# ============================================================================
# PATTERN 1: GLOBAL POLICY ENFORCEMENT
# ============================================================================

"""
Global middleware enforces a specific policy on all routes.
Agent ID is extracted from X-Agent-Passport-Id header.
"""
app.add_middleware(
    agent_passport_middleware,
    options=AgentPassportMiddlewareOptions(
        policy_id="finance.payment.refund.v1",  # Enforces refunds policy globally
        fail_closed=True
    )
)

# All routes below now require finance.payment.refund.v1 policy compliance
@app.post("/api/refunds")
async def process_refund(request: Request):
    """Process refund with global policy enforcement."""
    # Policy already verified - safe to process
    body = await request.json()
    amount = body.get("amount")
    currency = body.get("currency")
    order_id = body.get("order_id")
    
    return JSONResponse({
        "success": True,
        "refund_id": f"ref_{int(time.time() * 1000)}",
        "amount": amount,
        "currency": currency,
        "order_id": order_id,
        "agent_id": request.state.agent.agent_id
    })

# ============================================================================
# PATTERN 2: ROUTE-SPECIFIC WITH EXPLICIT AGENT ID (PREFERRED)
# ============================================================================

"""
Explicit agent ID is most secure and clear.
No header extraction needed.
"""
@app.post(
    "/api/data/export",
    dependencies=[Depends(require_policy("data.export.create.v1", AGENT_ID))],
)
async def export_data(request: Request):
    """Export data with explicit agent ID."""
    # Policy verified with explicit agent ID
    body = await request.json()
    rows = body.get("rows")
    format = body.get("format")
    contains_pii = body.get("contains_pii")
    
    return JSONResponse({
        "success": True,
        "export_id": f"exp_{int(time.time() * 1000)}",
        "rows": rows,
        "format": format,
        "contains_pii": contains_pii,
        "agent_id": request.state.agent.agent_id
    })

# ============================================================================
# PATTERN 3: ROUTE-SPECIFIC WITH HEADER FALLBACK
# ============================================================================

"""
Header fallback for backward compatibility.
Uses X-Agent-Passport-Id header.
"""
@app.post(
    "/api/messages/send",
    dependencies=[Depends(require_policy("messaging.message.send.v1"))],
)
async def send_message(request: Request):
    """Send message with header fallback."""
    # Policy verified via header
    body = await request.json()
    channel = body.get("channel")
    message_count = body.get("message_count")
    mentions = body.get("mentions")
    
    return JSONResponse({
        "success": True,
        "message_id": f"msg_{int(time.time() * 1000)}",
        "channel": channel,
        "message_count": message_count,
        "mentions": mentions,
        "agent_id": request.state.agent.agent_id
    })

# ============================================================================
# PATTERN 4: CUSTOM CONTEXT
# ============================================================================

"""
Custom context for complex scenarios.
"""
@app.post(
    "/api/repo/pr",
    dependencies=[
        Depends(
            require_policy_dependency_with_context(
                "code.repository.merge.v1",
                {
                    "action": "pr.create",
                    "branch": "main",
                    "repository": "myorg/myrepo",
                    "base_branch": "main"
                },
                AGENT_ID
            )
        )
    ],
)
async def create_pr(request: Request):
    """Create PR with custom context."""
    # Policy verified with custom context
    body = await request.json()
    pr_size_kb = body.get("pr_size_kb")
    file_path = body.get("file_path")
    
    return JSONResponse({
        "success": True,
        "pr_id": f"pr_{int(time.time() * 1000)}",
        "repository": "myorg/myrepo",
        "base_branch": "main",
        "pr_size_kb": pr_size_kb,
        "file_path": file_path,
        "agent_id": request.state.agent.agent_id
    })

# ============================================================================
# ERROR HANDLING
# ============================================================================

@app.exception_handler(HTTPException)
async def http_exception_handler(request: Request, exc: HTTPException):
    """Handle HTTP exceptions from middleware."""
    return JSONResponse(
        status_code=exc.status_code,
        content=exc.detail
    )

@app.exception_handler(Exception)
async def general_exception_handler(request: Request, exc: Exception):
    """Handle general exceptions."""
    return JSONResponse(
        status_code=500,
        content={
            "error": "internal_error",
            "message": "Internal server error"
        }
    )

# ============================================================================
# HEALTH CHECK
# ============================================================================

@app.get("/health")
async def health_check():
    """Health check endpoint."""
    return JSONResponse({
        "status": "ok",
        "timestamp": "2025-01-16T00:00:00Z",
        "agent_id": AGENT_ID
    })

# ============================================================================
# STARTUP
# ============================================================================

if __name__ == "__main__":
    import time
    
    print("🚀 Agent Passport FastAPI Example Server starting...")
    print(f"📋 Agent ID: {AGENT_ID}")
    print("\n📋 Test your endpoints:")
    print("\n1. Refunds (Global Policy):")
    print(f"curl -X POST 'http://localhost:8000/api/refunds' \\")
    print(f"  -H 'Content-Type: application/json' \\")
    print(f"  -H 'X-Agent-Passport-Id: {AGENT_ID}' \\")
    print("  -d '{\"amount\": 25.00, \"currency\": \"USD\", \"order_id\": \"order_123\", \"customer_id\": \"cust_456\", \"reason_code\": \"defective\", \"idempotency_key\": \"idem_789\"}'")
    
    print("\n2. Data Export (Explicit Agent ID):")
    print("curl -X POST 'http://localhost:8000/api/data/export' \\")
    print("  -H 'Content-Type: application/json' \\")
    print("  -d '{\"rows\": 1000, \"format\": \"json\", \"contains_pii\": false}'")
    
    print("\n3. Messaging (Header Fallback):")
    print(f"curl -X POST 'http://localhost:8000/api/messages/send' \\")
    print(f"  -H 'Content-Type: application/json' \\")
    print(f"  -H 'X-Agent-Passport-Id: {AGENT_ID}' \\")
    print("  -d '{\"channel\": \"general\", \"message_count\": 5, \"mentions\": [\"@user1\"]}'")
    
    print("\n4. Repository PR (Custom Context):")
    print("curl -X POST 'http://localhost:8000/api/repo/pr' \\")
    print("  -H 'Content-Type: application/json' \\")
    print("  -d '{\"pr_size_kb\": 50, \"file_path\": \"src/main.py\"}'")
    
    uvicorn.run(app, host="0.0.0.0", port=8000)

# ============================================================================
# SUMMARY: SIMPLE PATTERNS
# ============================================================================

"""
KEY PATTERNS:

✅ GLOBAL: Enforce one policy on all routes
   app.add_middleware(agent_passport_middleware, options=AgentPassportMiddlewareOptions(policy_id="finance.payment.refund.v1"))

✅ EXPLICIT: Most secure, explicit agent ID
   @app.post("/api/refunds", dependencies=[Depends(require_policy("finance.payment.refund.v1", AGENT_ID))])

✅ HEADER: Backward compatible, uses header
   @app.post("/api/export", dependencies=[Depends(require_policy("data.export.create.v1"))])

✅ CONTEXT: Custom context for complex scenarios
   @app.post("/api/pr", dependencies=[Depends(require_policy_dependency_with_context("code.repository.merge.v1", context, AGENT_ID))])

THAT'S IT! Simple, clear, and powerful policy enforcement.
"""
